File: python/src/middleware/requestsize.py
"""Request body size limit middleware.

# SECURITY LESSON: Without a body size limit, an attacker can send a multi-gigabyte request
# body to exhaust your server's memory. This is a trivial denial-of-service attack that
# costs the attacker almost nothing. Always cap request body size.
"""

import os

import structlog
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import JSONResponse

logger = structlog.get_logger()

# Default: 1 MB
MAX_REQUEST_BODY_BYTES = int(os.environ.get("MAX_REQUEST_BODY_BYTES", "1048576"))


class RequestSizeLimitMiddleware(BaseHTTPMiddleware):
    """Reject requests whose Content-Length exceeds the configured maximum."""

    def __init__(self, app, max_bytes: int = MAX_REQUEST_BODY_BYTES):
        super().__init__(app)
        self.max_bytes = max_bytes

    async def dispatch(self, request: Request, call_next):
        # SECURITY LESSON: Check Content-Length header before reading the body.
        # This rejects oversized requests early, before the server wastes memory buffering them.
        # Note: this doesn't protect against chunked transfer-encoding without Content-Length.
        # For full protection, also configure uvicorn's --limit-request-body or use a
        # reverse proxy (Traefik/nginx) with client_max_body_size.
        content_length = request.headers.get("content-length")

        if content_length is not None:
            try:
                length = int(content_length)
            except ValueError:
                # SECURITY LESSON: Malformed Content-Length headers can confuse proxies
                # and cause request smuggling. Reject them outright.
                return JSONResponse(
                    status_code=400,
                    content={"error": "Invalid Content-Length header"},
                )

            if length > self.max_bytes:
                logger.warning(
                    "request body too large",
                    content_length=length,
                    max_bytes=self.max_bytes,
                    path=request.url.path,
                    client=request.client.host if request.client else "unknown",
                )
                return JSONResponse(
                    status_code=413,
                    content={"error": "Request body too large"},
                    headers={
                        # SECURITY LESSON: Tell the client what the limit is so they can
                        # adjust. This is a courtesy, not a security requirement.
                        "X-Max-Content-Length": str(self.max_bytes),
                    },
                )

        return await call_next(request)

File: go/middleware/requestsize.go
package middleware

import (
	"log/slog"
	"net/http"
	"os"
	"strconv"
)

// SECURITY LESSON: Request size limits — Without a cap on request body size,
// an attacker can send a multi-gigabyte POST body and exhaust your server's
// memory (memory exhaustion / resource exhaustion attack). http.MaxBytesReader
// wraps the body and returns an error once the limit is exceeded, preventing
// the entire body from being read into memory. This is defense-in-depth:
// even if the reverse proxy has its own limit, the application should enforce
// its own to avoid depending on infrastructure configuration.

const (
	// defaultMaxRequestBytes is 1MB — sufficient for most JSON API payloads.
	defaultMaxRequestBytes int64 = 1 << 20
)

// MaxRequestSize returns middleware that limits the request body to maxBytes.
// If maxBytes is 0, it reads from the MAX_REQUEST_BODY_BYTES env var,
// defaulting to 1MB.
// Returns 413 Payload Too Large when the body exceeds the limit.
func MaxRequestSize(maxBytes int64, next http.Handler) http.Handler {
	if maxBytes <= 0 {
		maxBytes = defaultMaxRequestBytes
		if v := os.Getenv("MAX_REQUEST_BODY_BYTES"); v != "" {
			if parsed, err := strconv.ParseInt(v, 10, 64); err == nil && parsed > 0 {
				maxBytes = parsed
			}
		}
	}

	slog.Info("request size limit configured", "max_bytes", maxBytes)

	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		// SECURITY LESSON: http.MaxBytesReader is Go's built-in defense against
		// oversized request bodies. It wraps r.Body so that any read beyond the
		// limit returns an error. The http.Server then automatically returns
		// 413 if the handler tries to read beyond the limit. We also set it
		// explicitly here for handlers that read the body themselves.
		r.Body = http.MaxBytesReader(w, r.Body, maxBytes)

		next.ServeHTTP(w, r)
	})
}

// RequestSizeMiddleware is a convenience wrapper that reads the limit from
// the environment and applies it as middleware. Use this when wiring the
// handler chain in main.go.
func RequestSizeMiddleware(next http.Handler) http.Handler {
	return MaxRequestSize(0, next)
}
